src/transformStateWithClones.js

```diff
@@ -1,13 +1,62 @@
 'use strict';
 
+const ACTION_CASES = {
+  add: 'addProperties',
+  remove: 'removeProperties',
+  clear: 'clear',
+};
+
 /**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
+ * Deep clone function to create a copy of an object.
+ * @param {Object} obj - The object to be cloned.
+ * @returns {Object} - The deep clone of the object.
  */
+function simpleClone(obj) {
+  if (!obj || typeof obj !== 'object') {
+    return obj;
+  }
+
+  const objCopy = Array.isArray(obj) ? [] : {};
+
+  for (const key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      objCopy[key] = simpleClone(obj[key]);
+    }
+  }
+
+  return objCopy;
+}
+
+function transformState(currentStateCopy, action) {
```

Comment:
you can use destructuring here:

`const { type, extraData, keysToRemove } = action;`